addons/mail_plugin/tests/test_controller.py

```diff
@@ -0,0 +1,251 @@
+        # Test that we do not duplicate the company and that we return the previous one
+        mock_iap_enrich = Mock()
+        result = self._get_partner("Test", "qsd@" + domain, mock_iap_enrich)
+        self.assertFalse(
+            mock_iap_enrich.called,
+            "We already enriched this company, should not call IAP a second time")
+
+        second_company_id = result["partner"]["company"]["id"]
+        self.assertEqual(first_company_id, second_company_id, "Should not create a new company")
+
+        # But the same blacklisted domain on a different local part
+        # should create a new company (e.g.: asbl_XXXX@gmail.com VS asbl_YYYY@gmail.com)
+        result = self._get_partner(
+            "Test", "asbl@" + domain,
+            lambda _, domain: {"name": "Name", "email": ["asbl@" + domain]},
+        )
+        second_company_id = result["partner"]["company"]["id"]
+        self.assertNotEqual(first_company_id, second_company_id, "Should create a new company")
+
+    def test_get_partner_no_email_returned_by_iap(self):
+        """Test the case where IAP do not return an email address.
+
+        We should not duplicate the previously enriched company and we should be able to
+        retrieve the first one.
+        """
+        result = self._get_partner(
+            "Test", "qsd@domain.com",
+            lambda _, domain: {"name": "Name", "email": []},
+        )
+
+        first_company_id = result["partner"]["company"]["id"]
+        self.assertTrue(first_company_id and first_company_id > 0)
+
+        first_company = self.env["res.partner"].browse(first_company_id)
+        self.assertEqual(first_company.name, "Name")
+        self.assertFalse(first_company.email)
+
+        # Test that we do not duplicate the company and that we return the previous one
+        result = self._get_partner(
+            "Test", "qsd@domain.com",
+            lambda _, domain: {"name": "Name", "email": ["contact@" + domain]},
+        )
+        second_company_id = result["partner"]["company"]["id"]
+        self.assertEqual(first_company_id, second_company_id, "Should not create a new company")
+
+    def test_enrich_and_create_company(self):
+        partner = self.env["res.partner"].create({
+            "name": "Test partner",
+            "email": "test@test_domain.xyz",
+            "is_company": False,
+        })
+
+        result = self._enrich_and_create_company(
+            partner.id,
+            lambda _, domain: {"return": domain},
+        )
+
+        self.assertEqual(result["enrichment_info"], {"type": "company_created"})
+        self.assertEqual(result["company"]["additionalInfo"]["return"], "test_domain.xyz")
+
+        company_id = result["company"]["id"]
+        company = self.env["res.partner"].browse(company_id)
+        partner.invalidate_cache()
+        self.assertEqual(partner.parent_id, company, "Should change the company of the partner")
+
+    def _get_partner(self, name, email, patched_iap_enrich):
```

Comment:
I would extract it in a common.py so that it can be improved / reused easily (and probably call it mock_plugin_partner_get)